cpp/src/arrow/buffer.h

```diff
@@ -142,16 +147,18 @@ class ARROW_EXPORT Buffer {
 
   /// \brief Construct a new buffer that owns its memory from a std::string
   ///
+  /// The string data is copied into the newly-allocated buffer memory.
+  ///
   /// \param[in] data a std::string object
   /// \param[in] pool a memory pool
-  /// \param[out] out the created buffer
-  ///
-  /// \return Status message
+  static Result<std::shared_ptr<Buffer>> FromString(
+      const std::string& data, MemoryPool* pool ARROW_MEMORY_POOL_DEFAULT);
```

Comment:
This isn't directly relevant to this PR, but: why do we provide a version of this function which allocates from `pool`? We don't use std::string to store large quantities so it seems we could provide only

```c++
  static std::unique_ptr<Buffer> FromString(std::string data); 
```

In the case of a string which *does* contain large data, we can use:
```c++
ARROW_ASSIGN_OR_RAISE(auto buf, Buffer(large_string).Copy(pool));
```
Or so

Reply:
I don't know, though it doesn't hurt (the pool parameter is optional).

Reply:
But it still returns `Result<shared_ptr<Buffer>>` instead of just `shared_ptr<Buffer>`. The only reason to have the ability to fail is in case allocation fails, and I'm not sure that's a relevant case for things we store in a `std::string`

Reply:
cc @wesm for opinions.

Reply:
Follow up JIRA https://issues.apache.org/jira/browse/ARROW-8367